hydro_lang/src/keyed_stream.rs

```diff
@@ -75,7 +75,75 @@ where
     }
 }
 
+/// Helper function to create a KeyedStream from a transformed underlying stream,
+/// updating the metadata to reflect keyed stream type conversions.
+fn keyed_stream_with_updated_metadata<'a, K, U, L, B, O, R>(
+    underlying: Stream<(K, U), L, B, NoOrder, R>,
+) -> KeyedStream<K, U, L, B, O, R>
+where
+    L: Location<'a>,
+{
+    use crate::ir::{StreamKind, StreamOrdering, StreamRetries};
+
+    let ordering = match std::any::type_name::<O>() {
+        name if name.contains("TotalOrder") => StreamOrdering::TotalOrder,
+        _ => StreamOrdering::NoOrder,
+    };
+
+    let retries = match std::any::type_name::<R>() {
+        name if name.contains("ExactlyOnce") => StreamRetries::ExactlyOnce,
```

Comment:
This is awfully ugly code... We should define a `trait Ordering` and `trait Retries` that require an `fn kind() -> Stream{Ordering,Retries}` and use that.